cfg/builder/builder_finalize.cc

```diff
@@ -453,22 +453,80 @@ vector<UIntSet> CFGBuilder::fillInBlockArguments(core::Context ctx, const CFG::R
     return upperBounds1;
 }
 
-int CFGBuilder::topoSortFwd(vector<BasicBlock *> &target, int nextFree, BasicBlock *currentBB) {
-    // ENFORCE(!marked[currentBB]) // graph is cyclic!
-    if (currentBB->fwdId != -1) {
-        return nextFree;
-    } else {
-        currentBB->fwdId = -2;
-        if (currentBB->bexit.thenb->outerLoops >= currentBB->bexit.elseb->outerLoops) {
-            nextFree = topoSortFwd(target, nextFree, currentBB->bexit.elseb);
-            nextFree = topoSortFwd(target, nextFree, currentBB->bexit.thenb);
-        } else {
-            nextFree = topoSortFwd(target, nextFree, currentBB->bexit.thenb);
-            nextFree = topoSortFwd(target, nextFree, currentBB->bexit.elseb);
+namespace {
+
+// The marker for nodes that the post-order traversal has not yet encountered.
+constexpr int UNVISITED = -1;
+
+// The marker for nodes that the post-order traversal is processing the children of (this implies that the node is
+// somewhere in the work queue).
+constexpr int PROCESSING = -2;
+
+struct WorkItem {
+    enum class Event { Enter, Exit };
+
+    BasicBlock *block;
+    Event event;
+
+    explicit WorkItem(BasicBlock *block) : block{block}, event{Event::Enter} {}
+};
+
+} // namespace
+
+std::vector<int> CFGBuilder::topoSortFwd(vector<BasicBlock *> &target, int numBlocks, BasicBlock *entryBB) {
+    ENFORCE(target.empty(), "The output vector must be empty to start with");
+
+    target.reserve(numBlocks);
+
+    std::vector<int> forwardIndex(numBlocks, UNVISITED);
+    std::vector<WorkItem> work;
+
+    // Arbitrary, could use tuning
+    work.reserve(10);
+
+    work.emplace_back(entryBB);
+
+    while (!work.empty()) {
+        auto [block, event] = work.back();
+
+        switch (event) {
+            case WorkItem::Event::Enter: {
+                // Have we seen this node already through another path?
+                if (forwardIndex[block->id] != UNVISITED) {
+                    work.pop_back();
+                    continue;
+                }
+
+                // Ensure that we record this node on the way back out, mark it as visited, and enqueue its successors.
+                work.back().event = WorkItem::Event::Exit;
+
+                forwardIndex[block->id] = PROCESSING;
+
+                if (forwardIndex[block->bexit.thenb->id] == UNVISITED) {
+                    work.emplace_back(block->bexit.thenb);
+                }
+
+                if (!block->bexit.isUnconditional() && forwardIndex[block->bexit.elseb->id] == UNVISITED) {
+                    work.emplace_back(block->bexit.elseb);
+                }
```

Comment:
These two conditionals don't contribute to correctness, but do help to keep the depth of the work queue low. If they weren't present and we unconditionally added the entries to the queue, we'd get roughly the same behavior we see on the master branch: we would process them in a future iteration, notice that their index was not `UNVISITED` on line 495, pop the queue and continue working.